File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/fstree/capture.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package fstree

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/vmihailenco/msgpack/v5"
	"github.com/zeebo/blake3"
)

// Common errors
var (
	ErrTooManyFiles = errors.New("fstree: too many files")
	ErrFileTooLarge = errors.New("fstree: file too large")
	ErrCyclicLink   = errors.New("fstree: cyclic symbolic link detected")
)

// Capture takes a snapshot of the filesystem at the given root path.
// Returns a Snapshot containing the Merkle tree of all files and directories.
//
// The snapshot uses content-addressing:
//   - Unchanged files have the same hash across snapshots
//   - Unchanged directories have the same tree hash
//   - This enables efficient deduplication in the CXDB blob store
func Capture(root string, opts ...Option) (*Snapshot, error) {
	start := time.Now()

	// Resolve to absolute path
	absRoot, err := filepath.Abs(root)
	if err != nil {
		return nil, fmt.Errorf("resolve root: %w", err)
	}

	// Check root exists and is a directory
	info, err := os.Stat(absRoot)
	if err != nil {
		return nil, fmt.Errorf("stat root: %w", err)
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("root is not a directory: %s", absRoot)
	}

	// Apply options
	o := defaultOptions()
	for _, opt := range opts {
		opt(o)
	}

	// Build the tree
	b := &builder{
		root:     absRoot,
		opts:     o,
		trees:    make(map[[32]byte][]byte),
		files:    make(map[[32]byte]*FileRef),
		symlinks: make(map[[32]byte]string),
		visited:  make(map[string]bool), // for cycle detection with symlinks
	}

	rootHash, err := b.buildTree(absRoot, "")
	if err != nil {
		return nil, err
	}

	return &Snapshot{
		RootHash:   rootHash,
		Trees:      b.trees,
		Files:      b.files,
		Symlinks:   b.symlinks,
		CapturedAt: start,
		Stats: SnapshotStats{
			FileCount:    b.fileCount,
			DirCount:     b.dirCount,
			SymlinkCount: b.symlinkCount,
			TotalBytes:   b.totalBytes,
			Duration:     time.Since(start),
		},
	}, nil
}

// builder accumulates state during tree construction.
type builder struct {
	root     string
	opts     *options
	trees    map[[32]byte][]byte
	files    map[[32]byte]*FileRef
	symlinks map[[32]byte]string // target path for symlinks
	visited  map[string]bool     // resolved paths for cycle detection

	fileCount    int
	dirCount     int
	symlinkCount int
	totalBytes   uint64
}

// buildTree recursively builds the tree for a directory.
// Returns the hash of the TreeObject for this directory.
func (b *builder) buildTree(absPath, relPath string) ([32]byte, error) {
	// Check for cycles (when following symlinks)
	realPath, err := filepath.EvalSymlinks(absPath)
	if err == nil {
		if b.visited[realPath] {
			return [32]byte{}, ErrCyclicLink
		}
		b.visited[realPath] = true
		defer delete(b.visited, realPath)
	}

	// Read directory entries
	dirEntries, err := os.ReadDir(absPath)
	if err != nil {
		return [32]byte{}, fmt.Errorf("read dir %s: %w", relPath, err)
	}

	// Build entries for this directory
	var entries []TreeEntry

	for _, de := range dirEntries {
		name := de.Name()
		childRelPath := filepath.Join(relPath, name)
		childAbsPath := filepath.Join(absPath, name)

		// Check exclusions
		if b.opts.shouldExclude(childRelPath, de.IsDir()) {
			continue
		}

		// Get file info (follows symlinks if needed)
		var info fs.FileInfo
		if b.opts.followSymlinks {
			info, err = os.Stat(childAbsPath)
		} else {
			info, err = os.Lstat(childAbsPath)
		}
		if err != nil {
			// Skip files we can't stat (permission errors, etc.)
			continue
		}

		entry, err := b.buildEntry(childAbsPath, childRelPath, name, info)
		if err != nil {
			if errors.Is(err, ErrTooManyFiles) || errors.Is(err, ErrCyclicLink) {
				return [32]byte{}, err
			}
			// Skip individual files on error
			continue
		}

		entries = append(entries, entry)
	}

	// Sort entries by name for deterministic hashing
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Name < entries[j].Name
	})

	// Serialize and hash the tree object
	treeBytes, err := serializeTree(entries)
	if err != nil {
		return [32]byte{}, fmt.Errorf("serialize tree %s: %w", relPath, err)
	}

	hash := blake3.Sum256(treeBytes)
	b.trees[hash] = treeBytes
	b.dirCount++

	return hash, nil
}

// buildEntry creates a TreeEntry for a single filesystem entry.
func (b *builder) buildEntry(absPath, relPath, name string, info fs.FileInfo) (TreeEntry, error) {
	mode := uint32(info.Mode().Perm())

	switch {
	case info.Mode()&fs.ModeSymlink != 0:
		// Symbolic link - hash the target path
		target, err := os.Readlink(absPath)
		if err != nil {
			return TreeEntry{}, fmt.Errorf("readlink %s: %w", relPath, err)
		}

		hash := blake3.Sum256([]byte(target))
		b.symlinkCount++

		// Store symlink target string (not as FileRef since content is the target path)
		b.symlinks[hash] = target

		return TreeEntry{
			Name: name,
			Kind: EntryKindSymlink,
			Mode: mode,
			Size: uint64(len(target)),
			Hash: hash,
		}, nil

	case info.IsDir():
		// Directory - recurse
		dirHash, err := b.buildTree(absPath, relPath)
		if err != nil {
			return TreeEntry{}, err
		}

		return TreeEntry{
			Name: name,
			Kind: EntryKindDirectory,
			Mode: mode,
			Size: 0,
			Hash: dirHash,
		}, nil

	default:
		// Regular file
		if b.fileCount >= b.opts.maxFiles {
			return TreeEntry{}, ErrTooManyFiles
		}

		size := info.Size()
		if size > b.opts.maxFileSize {
			return TreeEntry{}, fmt.Errorf("%w: %s (%d bytes)", ErrFileTooLarge, relPath, size)
		}

		hash, err := hashFile(absPath)
		if err != nil {
			return TreeEntry{}, fmt.Errorf("hash file %s: %w", relPath, err)
		}

		b.files[hash] = &FileRef{
			Path: absPath,
			Size: uint64(size),
			Hash: hash,
		}
		b.fileCount++
		b.totalBytes += uint64(size)

		return TreeEntry{
			Name: name,
			Kind: EntryKindFile,
			Mode: mode,
			Size: uint64(size),
			Hash: hash,
		}, nil
	}
}

// hashFile computes the BLAKE3-256 hash of a file's contents.
func hashFile(path string) ([32]byte, error) {
	f, err := os.Open(path)
	if err != nil {
		return [32]byte{}, err
	}
	defer func() { _ = f.Close() }()

	h := blake3.New()
	if _, err := io.Copy(h, f); err != nil {
		return [32]byte{}, err
	}

	var hash [32]byte
	copy(hash[:], h.Sum(nil))
	return hash, nil
}

// serializeTree serializes a list of TreeEntry to msgpack.
// Uses numeric field tags matching the TreeEntry struct tags.
func serializeTree(entries []TreeEntry) ([]byte, error) {
	buf := &bytes.Buffer{}
	enc := msgpack.NewEncoder(buf)
	enc.SetSortMapKeys(true)

	if err := enc.Encode(entries); err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

// DeserializeTree deserializes msgpack bytes to TreeEntry slice.
func DeserializeTree(data []byte) ([]TreeEntry, error) {
	var entries []TreeEntry
	if err := msgpack.Unmarshal(data, &entries); err != nil {
		return nil, err
	}
	return entries, nil
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/fstree/capture.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime};

use blake3::Hasher;

use crate::encoding::encode_msgpack;

use super::options::{Options, SnapshotOption};
use super::types::{
    EntryKindDirectory, EntryKindFile, EntryKindSymlink, FileRef, Snapshot, SnapshotStats,
    TreeEntry,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FstreeErrorKind {
    TooManyFiles,
    FileTooLarge,
    CyclicLink,
    Io,
    Msgpack,
    Client,
    Other,
}

#[derive(Debug)]
pub struct FstreeError {
    pub kind: FstreeErrorKind,
    pub detail: String,
}

impl std::fmt::Display for FstreeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "fstree: {}", self.detail)
    }
}

impl std::error::Error for FstreeError {}

impl FstreeError {
    pub(crate) fn new(kind: FstreeErrorKind, detail: impl Into<String>) -> Self {
        Self {
            kind,
            detail: detail.into(),
        }
    }
}

#[allow(non_upper_case_globals)]
pub const ErrTooManyFiles: FstreeErrorKind = FstreeErrorKind::TooManyFiles;
#[allow(non_upper_case_globals)]
pub const ErrFileTooLarge: FstreeErrorKind = FstreeErrorKind::FileTooLarge;
#[allow(non_upper_case_globals)]
pub const ErrCyclicLink: FstreeErrorKind = FstreeErrorKind::CyclicLink;

pub type Result<T> = std::result::Result<T, FstreeError>;

pub fn capture(
    root: impl AsRef<Path>,
    opts: impl IntoIterator<Item = SnapshotOption>,
) -> Result<Snapshot> {
    let start = SystemTime::now();
    let abs_root = fs::canonicalize(root.as_ref())
        .map_err(|err| FstreeError::new(FstreeErrorKind::Io, err.to_string()))?;

    let metadata = fs::metadata(&abs_root)
        .map_err(|err| FstreeError::new(FstreeErrorKind::Io, err.to_string()))?;
    if !metadata.is_dir() {
        return Err(FstreeError::new(
            FstreeErrorKind::Other,
            format!("root is not a directory: {}", abs_root.display()),
        ));
    }

    let mut options = Options::default();
    for opt in opts {
        opt(&mut options);
    }

    let mut builder = Builder::new(options);
    let root_hash = builder.build_tree(&abs_root, Path::new(""))?;

    Ok(Snapshot {
        root_hash,
        trees: builder.trees,
        files: builder.files,
        symlinks: builder.symlinks,
        captured_at: start,
        stats: SnapshotStats {
            file_count: builder.file_count,
            dir_count: builder.dir_count,
            symlink_count: builder.symlink_count,
            total_bytes: builder.total_bytes,
            duration: start.elapsed().unwrap_or(Duration::from_secs(0)),
        },
    })
}

pub fn deserialize_tree(data: &[u8]) -> Result<Vec<TreeEntry>> {
    crate::encoding::decode_msgpack_into(data)
        .map_err(|err| FstreeError::new(FstreeErrorKind::Msgpack, err.to_string()))
}

struct Builder {
    options: Options,
    trees: HashMap<[u8; 32], Vec<u8>>,
    files: HashMap<[u8; 32], FileRef>,
    symlinks: HashMap<[u8; 32], String>,
    visited: HashSet<PathBuf>,
    file_count: usize,
    dir_count: usize,
    symlink_count: usize,
    total_bytes: u64,
}

impl Builder {
    fn new(options: Options) -> Self {
        Self {
            options,
            trees: HashMap::new(),
            files: HashMap::new(),
            symlinks: HashMap::new(),
            visited: HashSet::new(),
            file_count: 0,
            dir_count: 0,
            symlink_count: 0,
            total_bytes: 0,
        }
    }

    fn build_tree(&mut self, abs_path: &Path, rel_path: &Path) -> Result<[u8; 32]> {
        if let Ok(real_path) = fs::canonicalize(abs_path) {
            if self.visited.contains(&real_path) {
                return Err(FstreeError::new(
                    FstreeErrorKind::CyclicLink,
                    "cyclic symbolic link detected",
                ));
            }
            self.visited.insert(real_path.clone());
        }

        let mut entries = Vec::new();
        let dir_entries = fs::read_dir(abs_path)
            .map_err(|err| FstreeError::new(FstreeErrorKind::Io, err.to_string()))?;

        for entry in dir_entries {
            let entry = match entry {
                Ok(entry) => entry,
                Err(_) => continue,
            };
            let file_name = entry.file_name();
            let name = file_name.to_string_lossy().to_string();
            let child_rel = rel_path.join(&name);
            let child_abs = abs_path.join(&name);
            let rel_str = child_rel.to_string_lossy();

            if self.options.should_exclude(
                &rel_str,
                entry.file_type().map(|t| t.is_dir()).unwrap_or(false),
            ) {
                continue;
            }

            let metadata = if self.options.follow_symlinks {
                fs::metadata(&child_abs)
            } else {
                fs::symlink_metadata(&child_abs)
            };
            let metadata = match metadata {
                Ok(meta) => meta,
                Err(_) => continue,
            };

            match self.build_entry(&child_abs, &child_rel, &name, &metadata) {
                Ok(entry) => entries.push(entry),
                Err(err) => {
                    if err.kind == FstreeErrorKind::TooManyFiles
                        || err.kind == FstreeErrorKind::CyclicLink
                    {
                        return Err(err);
                    }
                    // Skip individual file errors
                    continue;
                }
            }
        }

        entries.sort_by(|a, b| a.name.cmp(&b.name));
        let tree_bytes = encode_msgpack(&entries)
            .map_err(|err| FstreeError::new(FstreeErrorKind::Msgpack, err.to_string()))?;
        let hash = blake3::hash(&tree_bytes);
        self.trees.insert(*hash.as_bytes(), tree_bytes);
        self.dir_count += 1;

        if let Ok(real_path) = fs::canonicalize(abs_path) {
            self.visited.remove(&real_path);
        }

        Ok(*hash.as_bytes())
    }

    fn build_entry(
        &mut self,
        abs_path: &Path,
        rel_path: &Path,
        name: &str,
        metadata: &fs::Metadata,
    ) -> Result<TreeEntry> {
        let mode = metadata.permissions().perm_mode() & 0o7777;

        if metadata.file_type().is_symlink() && !self.options.follow_symlinks {
            let target = fs::read_link(abs_path)
                .map_err(|err| FstreeError::new(FstreeErrorKind::Io, err.to_string()))?;
            let target_str = target.to_string_lossy().to_string();
            let hash = blake3::hash(target_str.as_bytes());
            self.symlink_count += 1;
            self.symlinks.insert(*hash.as_bytes(), target_str.clone());
            return Ok(TreeEntry {
                name: name.to_string(),
                kind: EntryKindSymlink,
                mode,
                size: target_str.len() as u64,
                hash: *hash.as_bytes(),
            });
        }

        if metadata.is_dir() {
            let dir_hash = self.build_tree(abs_path, rel_path)?;
            return Ok(TreeEntry {
                name: name.to_string(),
                kind: EntryKindDirectory,
                mode,
                size: 0,
                hash: dir_hash,
            });
        }

        if self.file_count >= self.options.max_files {
            return Err(FstreeError::new(
                FstreeErrorKind::TooManyFiles,
                "too many files",
            ));
        }

        let size = metadata.len();
        if size as i64 > self.options.max_file_size {
            return Err(FstreeError::new(
                FstreeErrorKind::FileTooLarge,
                format!("file too large: {} ({} bytes)", rel_path.display(), size),
            ));
        }

        let hash = hash_file(abs_path)
            .map_err(|err| FstreeError::new(FstreeErrorKind::Io, err.to_string()))?;
        self.files.insert(
            hash,
            FileRef {
                path: abs_path.to_path_buf(),
                size,
                hash,
            },
        );
        self.file_count += 1;
        self.total_bytes += size;

        Ok(TreeEntry {
            name: name.to_string(),
            kind: EntryKindFile,
            mode,
            size,
            hash,
        })
    }
}

fn hash_file(path: &Path) -> std::io::Result<[u8; 32]> {
    let mut file = fs::File::open(path)?;
    let mut hasher = Hasher::new();
    let mut buf = [0u8; 8192];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    let hash = hasher.finalize();
    Ok(*hash.as_bytes())
}

trait PermissionsExt {
    fn perm_mode(&self) -> u32;
}

impl PermissionsExt for fs::Permissions {
    #[cfg(unix)]
    fn perm_mode(&self) -> u32 {
        use std::os::unix::fs::PermissionsExt;
        self.mode()
    }

    #[cfg(not(unix))]
    fn perm_mode(&self) -> u32 {
        0
    }
}
